gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/VectoredIOImpl.java

```diff
@@ -0,0 +1,382 @@
+import org.apache.hadoop.fs.impl.CombinedFileRange;
+
+@VisibleForTesting
+public class VectoredIOImpl implements Closeable {
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+  private final BlockingQueue taskQueue = new LinkedBlockingQueue<Runnable>();
+  private final VectoredReadOptions vectoredReadOptions;
+  private final GhfsGlobalStorageStatistics storageStatistics;
+  private final FileSystem.Statistics statistics;
+  private ExecutorService boundedThreadPool;
+
+  public VectoredIOImpl(
+      VectoredReadOptions vectoredReadOptions,
+      GhfsGlobalStorageStatistics storageStatistics,
+      FileSystem.Statistics statistics) {
+    this.vectoredReadOptions = vectoredReadOptions;
+    // same fixedThreadPool executor, but provided a way to query task queue
+    this.boundedThreadPool =
+        new ThreadPoolExecutor(
+            vectoredReadOptions.getReadThreads(),
+            vectoredReadOptions.getReadThreads(),
+            0L,
+            TimeUnit.MILLISECONDS,
+            taskQueue,
+            new ThreadFactoryBuilder()
+                .setNameFormat("vectoredRead-range-pool-%d")
+                .setDaemon(true)
+                .build());
+    this.statistics = statistics;
+    this.storageStatistics = storageStatistics;
+  }
+
+  /**
+   * Reads data from Google Cloud Storage using vectored I/O operations.
+   *
+   * @param ranges List of file ranges to read.
+   * @param allocate Function to allocate ByteBuffer for reading.
+   * @throws IOException if an I/O error occurs.
+   */
+  /**
+   * Reads data from Google Cloud Storage using vectored I/O operations.
+   *
+   * @param ranges List of file ranges to read.
+   * @param allocate Function to allocate ByteBuffer for reading.
+   * @param gcsFs GCFS implementation to use while creating channel and reading content for ranges.
+   * @param fileInfo FileInfo of the gcs object agaisnt which range request are fired, this can be
+   *     null for some code path fall back to URI path provided.
+   * @param gcsPath URI of the gcs object for which the range requests are fired.
+   * @throws IOException If invalid range is requested, offset<0.
+   */
+  public void readVectored(
+      List<? extends FileRange> ranges,
+      IntFunction<ByteBuffer> allocate,
+      GoogleCloudStorageFileSystem gcsFs,
+      FileInfo fileInfo,
+      @Nonnull URI gcsPath)
+      throws IOException {
+    ReadChannelProvider channelProvider = new ReadChannelProvider(gcsFs, fileInfo, gcsPath);
+    readVectored(ranges, allocate, channelProvider);
+  }
+
+  private void readVectored(
+      List<? extends FileRange> ranges,
+      IntFunction<ByteBuffer> allocate,
+      ReadChannelProvider channelProvider)
+      throws IOException {
+    List<? extends FileRange> sortedRanges = validateNonOverlappingAndReturnSortedRanges(ranges);
+    for (FileRange range : ranges) {
+      // TODO(user): upgrade to use validateAndSortRanges
```

Comment:
what is the advantage?

Reply:
It's the new method which will be part of new hadoop release. It's intutive from the name that it merges validation and sorting in one step.